draft-ietf-quic-transport.md

```diff
@@ -1528,31 +1560,44 @@ being used for denial of service.  See {{migrate-validate}} for details.
-{{packet-number-gap}}. Packet number gaps are cumulative.  A client might skip
-connection IDs, but it MUST ensure that it applies the associated packet number
-gaps for connection IDs that it skips in addition to the packet number gap
-associated with the connection ID that it does use.
-
-A server that receives a packet that is marked with a new connection ID recovers
-the packet number by adding the cumulative packet number gap to its expected
-packet number.  A server SHOULD discard packets that contain a smaller gap than
-it advertised.
+to correlate activity between those paths.  An endpoint that moves between
+networks might not wish to have their activity correlated by any entity other
+than a server. The NEW_CONNECTION_ID message can be sent by both endpoints to
+provide an unlinkable connection ID for use in case a peer wishes to explicitly
+break linkability between two points of network attachment.
+
+An endpoint might need to send packets on multiple networks without receiving
+any response from its peer.  To ensure that the endpoint is not linkable
+across each of these changes, a new connection ID and packet number gap are
+needed for each network.  To support this, each endpoint sends multiple
+NEW_CONNECTION_ID messages.  Each NEW_CONNECTION_ID is marked with a sequence
+number.  Connection IDs MUST be used in the order in which they are numbered.
+
+An endpoint that does not require the use of a connection ID should not request
+that its peer use a connection ID.  Such an endpoint does not need to provide
+new connection IDs using the NEW_CONNECTION_ID frame.
+
+An endpoint which wishes to break linkability upon changing networks MUST use
+the connection ID provided by its peer as well as incrementing the packet
+sequence number by an externally unpredictable value computed as described in
+{{packet-number-gap}}. Packet number gaps are cumulative.  An endpoint might
+skip connection IDs, but it MUST ensure that it applies the associated packet
+number gaps for connection IDs that it skips in addition to the packet number
+gap associated with the connection ID that it does use.
+
+An endpoint that receives a packet that is marked with a new connection ID
+recovers the packet number by adding the cumulative packet number gap to its
+expected packet number.  An endpoint MUST discard packets that contain a smaller
+gap than it advertised.
+
+An endpoint that receives a successfully authenticated packet with a previously
+unused connection ID MUST use the next available connection ID for any packets
+it sends to that address.  To avoid changing connection IDs multiple times when
+packets arrive out of order, endpoints MUST change only in response to a packet
+that increases the largest received packet number.  Failing to do this could
+allow for use of that connection ID to link activity on new paths.  There is no
+need to move to a new connection ID if the address of a peer changes without
+also changing the connection ID.
 
```

Comment:
I'm not fully up to speed on connection migration, so bear with me, but...

What if both endpoints migrate? For example B has sent two NEW_ID_B(1,2) to A, and A sends two NEW_ID_A(1,2) to B. A migrates and consumes the first NEW_ID_B(1). Then B migrates and consumes NEW_ID_A(1). But because either end has migrated, they are now on new infrastructure and the load balancers no longer understand NEW_ID_A(2) or NEW_ID_B(2).

It seems to me that you cannot have a sequential NEW_ID - assuming an endpoint only migrates to one location at a time (no multipath) there must a be a latest NEW_ID that will work, but the older might not. It is just trial an error until something works?

Reply:
Migration happens with short headers, so there is usually one connection ID in each direction.

For instance, the client might be sending SCID3 and the server might send CCID2.  If the client moves, then it sends SCID4.  The server sees the new connection ID from the client and moves to sending CCID3.

Note that connection IDs are currently strictly sequential, but that is not a property we need to preserve if we encrypt packet numbers.  Any connection ID should work (with some caveats that I can go into more).

Reply:
Hmm - in the above example: If the server later moves, it uses CCID4 and the client sees this and sends SCID5. But SCID5 was send in a NEW_ID message before the server moved to use CCID4 and the server is now behind infrastructure that does not understand SCID5.

If the migration where the server sends CCID4 also marks the smallest valid SCID, there should be no issue - perhaps this is already the case.

Reply:
It is the responsibility of the endpoint to ensure that any connection ID it provides to its peer is capable of reaching it, no matter where it moves to.  The only stipulation being that packets arrive at an address that the endpoint control.  Either the endpoint or intermediaries over which it has influence, like a load balancer.